messenger/make.py

```diff
@@ -67,36 +67,35 @@ def build_octave():
     do_build(make_cmd, 'messenger.mex')
 
 
-def build_matlab():
+def build_matlab(static=False):
     cfg = get_config()
-    matlab_bin = cfg['matlab_bin']
+    matlab_bin = cfg['matlab_bin'].strip('"')
     # Get the extension
-    if sys.platform == 'win32':
-        extcmd = '"%s\\mexext.bat"' % matlab_bin
-    else:
-        extcmd = matlab_bin + "/mexext"
-
-    check_extension = subprocess.Popen(extcmd, stdout=subprocess.PIPE)
-    extension = check_extension.stdout.read()
+    extcmd = '"' + os.path.join(matlab_bin, "mexext") + '"'
+    extension = subprocess.check_output(extcmd, shell=True)
     extension = extension.decode('utf-8').rstrip('\r\n')
 
     # Build the mex file
-    if sys.platform == 'win32':
-        mex = matlab_bin + "\\mex.bat"
-    else:
-        mex = matlab_bin + "/mex"
+    mex = '"' + os.path.join(matlab_bin, "mex") + '"'
     paths = "-L%(zmq_lib)s -I%(zmq_inc)s" % cfg
     make_cmd = '%s -O %s -lzmq ./src/messenger.c' % (mex, paths)
+    if static:
+        make_cmd += ' -DZMQ_STATIC'
     do_build(make_cmd, 'messenger.%s' % extension)
 
 
 if __name__ == '__main__':
-    usage = 'Please specify a valid make target (Matlab or Octave)'
-    if len(sys.argv) < 2:
-        print(usage)
-    elif sys.argv[1].lower() == 'matlab':
-        build_matlab()
-    elif sys.argv[1].lower() == 'octave':
-        build_octave()
+    import argparse
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "target",
+        choices=["matlab", "octave"],
+        type=str.lower,
+        help="target to be built")
+    parser.add_argument("--static", action="store_true",
+                        help="staticly link libzmq")
+    args = parser.parse_args()
+    if args.target == "matlab":
+        build_matlab(static=args.static)
     else:
-        print(usage)
+        build_octave()
```

Comment:
It won't let you choose another one because argparse limits the choices.

for example, running ./make.py foo will give:
usage: make.py [-h] [--static] {matlab,octave}
make.py: error: argument target: invalid choice: 'foo' (choose from
'matlab', 'octave')



On Mon, Mar 9, 2015 at 7:54 PM, Ariel Rokem <notifications@github.com>
wrote:

> In messenger/make.py
> <https://github.com/arokem/python-matlab-bridge/pull/161#discussion_r26095797>
> :
>
> >      else:
> > -        print(usage)
> > +        build_octave()
>
> This seems like it might be confusing. If someone enters a typo, it will
> try to build the octave thing and emit a confusing error. How about:
>
> elif args.target == 'octave':
>     build_octave
> else:
>     e_s = "'%s' is not a valid target for building the messenger."%args.target
>     e_s = e_s + " Use 'matlab' or 'octave'"
>     raise ValueError(e_s)
>
> Or something to that effect. Also, could you please rebase on master?
> Thanks!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/arokem/python-matlab-bridge/pull/161/files#r26095797>.
>